app/src/main/java/org/listenbrainz/android/model/ResponseError.kt

```diff
@@ -0,0 +1,58 @@
+package org.listenbrainz.android.model
+
+import com.google.gson.Gson
+import com.google.gson.reflect.TypeToken
+import retrofit2.Response
+
+/** These exceptions need to handled in view-model, shown via UI and not just thrown.
+ * @param genericToast Generic message for the error.
+ * @param actualResponse Actual response given by the server.*/
+interface ResponseError {
+    
+    val genericToast: String
+    var actualResponse: String?
+    
+    /** Simple function that returns the most suitable message to show the user.*/
+    fun toast(): String = actualResponse ?: genericToast
+    
+    companion object {
+        
+        /** Get [ResponseError] for a given Retrofit **error** [Response] from server.*/
+        fun <T> getSocialResponseError(response: Response<T>) : ResponseError {
+            val error = parseError(response)
+            return getSocialErrorType(error)
+        }
+        
+        /** Parsing server response into [ApiError] class. Consider using specific functions like [getSocialResponseError], etc. for each repository if
+         * returning errors is the sole motive.*/
+        fun <T> parseError(response: Response<T>) : ApiError =
+            Gson().fromJson(
+                /* json = */ response.errorBody()?.string(),
+                /* typeOfT = */ object : TypeToken<ApiError>() {}.type
+            )
+    
+    
+        /** Get [ResponseError] for social type API endpoints. Automatically puts actual error message.
+         * Prefer using [getSocialResponseError] in repository functions.*/
+        private fun getSocialErrorType(apiError: ApiError) : ResponseError {
+            val error = apiError.error
+            val code = apiError.code
+            return when {
+                    code == 404 -> SocialError.USER_NOT_FOUND
+                    error?.substringAfter(' ')?.contains("is already") == true -> SocialError.ALREADY_FOLLOWING
+                    error?.substringAfter(' ') == "cannot follow yourself." -> SocialError.CANNOT_FOLLOW_SELF
```

Comment:
Oh god. What is happening here?
You have gone too deep here and the whole logic will start breaking if the text is changed. Remove this asap. We should have defined error codes for these scenarios.

Reply:
Unfortunately we don't have unique error codes for these situations. Yes the code will break if something was to be changed but that's just how it is. Here I've tried to make it efficient so as to not compare the whole string. If somehow, the response is to change, UNKNOWN_ERROR will be thrown which is intended behaviour.

Reply:
The only way to identify the error is by the string in the response.
Maybe we can prompt lucifer to add an extra field so error detection is easy?
@amCap1712

Reply:
Why not just display the error message directly to the user in case of `code == 400`?

Reply:
@amCap1712 it restricts some amount of flexibility imo.

Reply:
Welcome to the land of overengineering